### PR TITLE
NONE: fix OOM

### PR DESCRIPTION
**Pull request by bgvozdev**
**What's in this PR?**
Fixing OOM

**Why**
Cause OOMs are bad, m'kay.

**Added feature flags**
Relying on tests and typescript

**How has this been tested?**  
tests and typescript

**Whats Next?**
Celerbrate! 🍾

### CODE REVIEW

File: src/models/reposyncstate.ts [hunk 1]
```diff
@@ -215,15 +215,26 @@ export class RepoSyncState extends Model implements RepoSyncStateProperties {
...
-			}
+			},
+			limit,
+			offset,
+			order
 		}));
 		return result || [];
 	}
 
+	static async countAllFromSubscription(subscription: Subscription, options: FindOptions = {}): Promise<number> {
```
Comment (atraversatlassian) [inline src/models/reposyncstate.ts:230]: This function looks to do the same thing as countSubscriptionRepos on line 182

File: src/routes/api/api-reset-subscription-failed-tasks.ts [hunk 2]
```diff
@@ -26,20 +26,27 @@ export const ApiResetSubscriptionFailedTasks = async (req: Request, res: Respons
...
+	let hasNextPage = true;
+	do {
+		const repoSyncStates = await RepoSyncState.findAllFromSubscription(subscription, 100, offset, [["repoFullName", "ASC"]]);
+		offset += 100;
+		hasNextPage = repoSyncStates.length === 100;
+
+		await Promise.all(repoSyncStates.map(async (repoSyncState) => {
+			let updated = false;
+			targetTasks.forEach(targetTask => {
+				if (repoSyncState[`${targetTask}Status`] === "failed") {
```
Comment (atraversatlassian) [inline src/routes/api/api-reset-subscription-failed-tasks.ts:39]: Could we pass this as filter to the RepoSyncState query and have it only return items that are failed instead?

File: src/routes/api/api-reset-subscription-failed-tasks.ts [hunk 3]
```diff
@@ -26,20 +26,27 @@ export const ApiResetSubscriptionFailedTasks = async (req: Request, res: Respons
...
-		let updated = false;
-		targetTasks.forEach(targetTask => {
-			if (repoSyncState[`${targetTask}Status`] === "failed") {
-				repoSyncState[`${targetTask}Status`] = null;
-				logs.push({ repoSyncStateId: repoSyncState.id, targetTask });
-				updated = true;
+	let offset = 0;
+	let hasNextPage = true;
+	do {
+		const repoSyncStates = await RepoSyncState.findAllFromSubscription(subscription, 100, offset, [["repoFullName", "ASC"]]);
```
Comment (atraversatlassian) [inline src/routes/api/api-reset-subscription-failed-tasks.ts:32]: nit would be better if the 100 value could come from config

File: src/routes/api/installation/api-installation-syncstate-get.ts [hunk 4]
```diff
@@ -4,9 +4,32 @@ import { Subscription } from "models/subscription";
...
+		return;
+	}
+
 	const githubInstallationId = Number(req.params.installationId);
 	const gitHubAppId = Number(req.params.gitHubAppId) || undefined;
 	const jiraHost = req.params.jiraHost;
+	const limit = Number(req.query.limit);
+	const offset = Number(req.query.offset);
+
+	if (limit > 10000) {
```
Comment (atraversatlassian) [inline src/routes/api/installation/api-installation-syncstate-get.ts:27]: nit 10000 coming from config better.  At least make it a constant value

**Review by atraversatlassian — DISMISSED** [review]

File: src/routes/api/api-reset-subscription-failed-tasks.ts (hunk 3 above)
Comment (bgvozdev) [reply src/routes/api/api-reset-subscription-failed-tasks.ts:32]: that's some API endpoint that we probably never use; I wouldn't bother too much about it

File: src/models/reposyncstate.ts (hunk 1 above)
Comment (bgvozdev) [reply src/models/reposyncstate.ts:230]: Awesome find! :killwithfire:

File: src/routes/api/api-reset-subscription-failed-tasks.ts (hunk 2 above)
Comment (bgvozdev) [reply src/routes/api/api-reset-subscription-failed-tasks.ts:39]: That was an existing code, don't want to touch it. I only wrapped it in a loop for pagination (changed how we call `findAllFromSubsscription`), the rest is left as it was.

File: src/routes/api/installation/api-installation-syncstate-get.ts (hunk 4 above)
Comment (bgvozdev) [reply src/routes/api/installation/api-installation-syncstate-get.ts:27]: just an admin endpoint we probably never used, I woudn't bother much

File: src/routes/api/api-reset-subscription-failed-tasks.ts (hunk 3 above)
Comment (bgvozdev) [reply src/routes/api/api-reset-subscription-failed-tasks.ts:32]: turned into a constant though :) thanks!

File: src/routes/api/installation/api-installation-syncstate-get.ts (hunk 4 above)
Comment (bgvozdev) [reply src/routes/api/installation/api-installation-syncstate-get.ts:27]: turned into a constant though :) thanks!

File: src/routes/api/installation/api-installation-syncstate-get.ts [hunk 5]
```diff
@@ -4,9 +4,34 @@ import { Subscription } from "models/subscription";
 import { pick } from "lodash";
 
 export const ApiInstallationSyncstateGet = async (req: Request, res: Response): Promise<void> => {
+	if (req.query.limit === undefined) {
```
Comment (atraversatlassian) [inline src/routes/api/installation/api-installation-syncstate-get.ts:7]: could query be undefined? and need a null propagator

**Review by atraversatlassian — APPROVED** [review]

File: src/routes/api/installation/api-installation-syncstate-get.ts (hunk 5 above)
Comment (bgvozdev) [reply src/routes/api/installation/api-installation-syncstate-get.ts:7]: No, I think it is always defined, at least according to express typing

**Pull request merged**